pkg/dio/multi_test.go

```diff
@@ -12,365 +12,278 @@
+	// *Added ones will be added to multi writer later
+	for i := 0; i < count; i++ {
+		var buffer bytes.Buffer
+
+		reader, writer := io.Pipe()
+
+		// add the writer & buffer to skip list if i divisible by three
+		if i%3 == 0 {
+			writersAdded = append(writersAdded, writer)
+			buffersAdded = append(buffersAdded, &buffer)
+		} else {
+			writers = append(writers, writer)
+			buffers = append(buffers, &buffer)
+		}
 
-	data2 := "verify dynamic add"
-	_, err = mwriter.Write([]byte(data2))
-	if err != nil {
-		t.Error(err)
-		return
+		// set up a goroutine so we don't block writes
+		go io.Copy(&buffer, reader)
 	}
 
-	// compare the data - shouldn't be present
-	if bufA.String() != data+data2 {
-		t.Errorf("A: expected: %s, actual: %s", data+data2, bufA.String())
-		return
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	if bufB.String() != data2 {
-		t.Errorf("B: expected: %s, actual: %s", data2, bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
 
+	// add skipped writers to the writer
+	mwriter.Add(writersAdded...)
+
+	write(t, mwriter, []byte(dynamic))
+	control(t, buffers, []byte(base+dynamic))
+	control(t, buffersAdded, []byte(dynamic))
 }
 
+// TestMultiWrite creates multi writers and writes to them, then removes some of them and writes again
 func TestWriteRemove(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+	var writers []io.Writer
+	var buffers []*bytes.Buffer
 
-	mwriter := MultiWriter(pipeAW, pipeBW)
+	var buffersLeft []*bytes.Buffer
+	var buffersRemoved []*bytes.Buffer
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
+	// create & initialize writers and buffers into two categories
+	// *Removed ones will be filtered out from multi writer later
+	for i := 0; i < count; i++ {
+		var buffer bytes.Buffer
 
-	// set up a copy so we don't block writes
-	go io.Copy(&bufA, pipeAR)
-	go io.Copy(&bufB, pipeBR)
+		reader, writer := io.Pipe()
 
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
-	}
+		writers = append(writers, writer)
+		buffers = append(buffers, &buffer)
 
-	// compare the data
-	if bufA.String() != data {
-		t.Errorf("A: expected: %s, actual: %s", data, bufA.String())
-		return
+		// set up a goroutine so we don't block writes
+		go io.Copy(&buffer, reader)
 	}
 
-	if bufB.String() != data {
-		t.Errorf("B: expected: %s, actual: %s", data, bufB.String())
-		return
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	// Remove the writer from the MultiWriter
-	mwriter.Remove(pipeBW)
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
 
-	data2 := "verify dynamic remove"
-	_, err = mwriter.Write([]byte(data2))
-	if err != nil {
-		t.Error(err)
-		return
-	}
+	// add the writer & buffer to skip list if i divisible by three
+	for i := 0; i < count; i++ {
```

Comment:
combine this with the preceding loop so the structure is the same as the Add test.